### ISSUE TITLE
Add a render pipeline to WebGLRenderer

### ISSUE THREAD

**Issue by benaadams** [mrdoob/three.js#6464]
From #6383 [We Need to Properly Handle Gamma Correction](https://github.com/mrdoob/three.js/issues/6383) and elements from [EffectComposer](https://github.com/mrdoob/three.js/blob/master/examples/js/postprocessing/EffectComposer.js) and [WebGLDeferredRenderer](https://github.com/mrdoob/three.js/blob/master/examples/js/renderers/WebGLDeferredRenderer.js) and inspired a little from [ASP.NET v5](https://github.com/aspnet/Mvc)'s modular HTTP request pipeline.

> Yes, we need to support tone mapping for scenes with HDR. Presumably, we need FXAA, too. And I am pushing for GammaCorrection. And users may want any number of post-processing effects.

To try to capture the varied activities, rather than the EffectCompser approach which is a one way queue style processor, I would suggest a two way pipeline where each stage has a chance to handle before and after (more on this later).

As example, the current default render pipeline would look something like the following:

``` javascript
pipeline.append( clearFrameBufferStage, THREE.InitalizeStage );
pipeline.append( uploadBuffersStage, THREE.InitalizeStage );
pipeline.append( uploadTexturesStage, THREE.InitalizeStage );
pipeline.append( compileProgramsStage, THREE.InitalizeStage );
pipeline.append( updateMatricesStage, THREE.InitalizeStage );
pipeline.append( sortRenderablesStage, THREE.InitalizeStage );
pipeline.append( renderShadowMapStage, THREE.PreRenderStage );
pipeline.append( renderOpaqueObjectsStage, THREE.RenderStage );
pipeline.append( renderSkyBoxStage, THREE.RenderStage );
pipeline.append( renderTransparentObjectsStage, THREE.RenderStage );
pipeline.append( renderLensFlareStage, THREE.PostRenderStage );
pipeline.append( gammaToneMapStage, THREE.FinalizeStage );
```

The signature of a stage would be the following and would be passed in the next stage to run as a parameter as well as the current render context/state (scene, framebuffers etc):

``` javascript
function stageName( context, next ) {

    ... // code to run at start

    next();

    ... // code to run at end

}
```

As the stage is passed in the next item in the pipeline, it can decide to do both before and after the next stage runs. 

So if you had a pipeline item that updated buffers or textures, if you wanted to do it at the start of the frame you'd do:

``` javascript
function uploadBuffersStage( context, next ) {

    objects.update( context.opaqueObjects );
    objects.update( context.transparentObjects );
    next();

}
```

But if you wanted to do it at end of frame post render you'd do:

``` javascript
function uploadBuffersStage( context, next ) {

    next();
    objects.update( context.opaqueObjects );
    objects.update( context.transparentObjects );

}
```

The stage markers are to make it easier to insert items either at the start or end of a stage from outside the renderer.

For example if you wanted to add a depth renderer before the objects were rendered, but not disturb the pipeline, you could do:

``` javascript
renderer.pipeline.prepend( renderDepthStage, THREE.RenderStage );
```

Or a lighting stage after the objects were rendered:

``` javascript
renderer.pipeline.append( renderLightingStage, THREE.RenderStage );
```

Or FXAA and DoF/BokehStage just before Gamma correction:

``` javascript
renderer.pipeline.prepend( postBokehStage, THREE.FinalizeStage );
renderer.pipeline.prepend( postAntiAliasFXAAStage, THREE.FinalizeStage );
```

Thoughts?

**Comment by benaadams** [comment]
For a better example of using both directions of the pipe something like lens flare might be at an earlier stage:
````javascript
pipeline.append( renderLensFlareStage, THREE.PreRenderStage );
````
And the stage might be as follows (though probably not how lens flare works, but more for the idea)
````javascript
function renderLensFlareStage(context, next) {
    
    renderOcculderProbe();
    next();
    renderFlares();

}
````
Or perhaps a post processing effect that needs to swap render targets
````javascript
function postProcess(context, next) {
    
    doFunkyEffect( context );
    context.swapRenderTargets( ); // swap read and write targets
    next();
    context.swapRenderTargets(); // swap read and write back

}
````

**Comment by gero3** [comment]
I don't dislike the idea but start and end should be either seperate fucntion inside a stage or should be be seperate stages. Becuase otherwise it would be hard to follow the flow of the code.

**Comment by mrdoob** [comment]
Hmmm, I guess that's the direction `WebGLRenderer` is going. It still needs more modularisation though.

**Comment by benaadams** [comment]
@gero3 more like?
```javascript
var stage = {
    execute: function ( context ) { },
    unwind: function ( context ) { }
}
```
@mrdoob more modularisation?

**Comment by mrdoob** [comment]
`WebGLRenderer` is slowly starting to be more readable and we can starting to see the "stages". We still need to move more stuff out of it (Textures, RenderTargets, ...) and then doing something like what you propose will become more easy/obvious.

**Comment by benaadams** [comment]
So maybe pause on this for a bit?

**Comment by mrdoob** [comment]
Yes please 😊

**Comment by bhouston** [comment]
Something in this direction would be nice.  I agree with @mrdoob to not yet move towards implementation.

My thoughts are based on the previous renderers I wrote, is that one can create modules to handle some things.  Such as Buffers, Textures and Programs.

I have had a BufferCache object that is responsible for uploading buffers and managing them as a pool.

I've had a TextureCache that is responsible for uploading textures and allow them to be shared easily.

I've had a ProgramCache as well.

So for resources that are shared between frames and objects, I prefer cache pattern.  Where you make requests and they get fulfilled.  And you can release them.  And you have central counts and usage statistics, etc.

But the best reason for having these central cache -- and it is very important -- if you lose context, you can ask all your caches to recreate their GPU resources.

It also moves these types of resource managers outside of the linear render pipeline -- because they do not have to be linearized.  Your program/texture/buffer caches can return promises and other types of dynamic things.

So I love the idea of stages and I do recommend them, but only for preRender, render and post render effects.

**Comment by tschw** [comment]
In order to move forward on this issue, it might be a good idea to specify a data model for the rendering system because
- it's a useful basis to sketch out the exact requirements,
- it's generally a good starting point for an implementation (most programmers will intuitively agree on that one and it's even "state of the art", if I remember Mr. Knuth correctly),
- persistence and editing won't be forgotten,
- formats like glTF are already encorporating metadata for that purpose to some degree.

As shaders can be used for all sorts of things, I would like to raise the question whether a modular rendering system should use a hardcoded pipeline. I think it is actually a dependency graph by nature. Example: Pass A goes to a rendertarget and pass b requires its texture, inducing a dependency - maybe a delay or rate divider of a certain number of frames is granted.
There can be a linearization at some point, but would a bottom-up approach as proposed here turn out flexible enough?

Another interesting questions is, whether shaders should always be GLSL monoliths or whether they should be snippets. In case of the latter we have have the same mechanics going on (this time at shader concatenation time to determine the order of snippets): Dependencies inferred from data connections.

Next question that arises is: What exactly is a material? Just a mapping of properties? Or can it plug passes into the pipeline and GLSL into the shaders? If not, what does? How do we call it? And how to get it sufficiently modularized? However, the interface definition would be basicallly yet another use case of a central property system (as roughly sketched out off-topic in #6557).

Thinking this stuff all through -I don't claim I have :grinning:- should boil down to just a few basic building blocks for a fully modular rendering system. In case we can sketch out something sound, I'd happily help pulling it off implementation-wise.

There are some very interesting points in the previous post:

> But the best reason for having these central cache -- and it is very important -- if you lose context, you
> can ask all your caches to recreate their GPU resources.

Yes. In abstract, it's pretty much the only sane way to handle context loss. Also, it shouldn't keep JS-side copies by default and instead require new input. There's caching in the Browser and caching extensive preprocessing should be in the hands of the user.

> It also moves these types of resource managers outside of the linear render pipeline -- because they do
> not have to be linearized. Your program/texture/buffer caches can return promises and other types of
> dynamic things.

But should this "cache" really be so distinct from the pipeline? And doesn't context loss actually have the same consequences as a changed input of a loader, generator or rendering pass? What becomes outdated needs to be rebuilt, right? 

For an animated scene, the time input and/or change of matrices invalidate the rendertargets, finally resulting in an updated framebuffer.

I think it would be a desirable goal to treat asynchronous tasks as first class citizens of such as system. Loading is inherently asynchronous and tasks that can go without a GL context (decompression, parsing, modeling operations that update buffers) may employ workers.

How to linearize should be a different concern. Futures (not to say "promises and other types of dynamic things") are the tool to tackle asynchronicity in this context. Also, there can be "hard" (retry later) and "soft" (proceed without) dependencies. Actually, there should be three levels of hardness as there are in fact three states: Uninitialized, outdated, and up-to-date. Note that before mentioned frame delay and rate dividers are a very similar concept in the synchronous world.
